### PR TITLE
[runtime] Use MAP_JIT automatically when running under a hardened runtime on osx.

### PR DESCRIPTION
**Pull request by vargaz**
This is to address https://github.com/mono/mono/pull/13446#issuecomment-489749599

### CODE REVIEW

**Comment by duncanmak** [comment]
I had one thought - maybe we should only probe if `use_map_jit` is not already explicitly turned on.

If we rearrange a bit, maybe the test becomes ``if ((flags & MONO_MMAP_JIT) || use_mmap_jit || is_hardened_runtime)`.

**Comment by jaykrell** [comment]
Or use a pthread_once which is for this kind of thing, except you also need to be coop-aware, perhaps (this early, before any JIT, maybe no matter).

**Comment by vargaz** [comment]
@monojenkins build failed

**Comment by jaykrell** [comment]
There is no restriction on the ordering of the reads. You can see the uninitialized ‘is_hardened_runtime’ and then ‘is_checked’ true. 
A trivalue integer works — uninitialized, ‘true’, ‘false’, such as 0, 1, 2 or such.
Ordering the writes does not prevent that, does not
Order the reads.
Barriers come in pairs, is the unobvious rule of thumb. 
I’ll dig up the explanation from the compiler developer.

**Comment by jaykrell** [comment]
https://lwn.net/Articles/576486/

“Memory barriers almost invariably come in pairs. If one of two cooperating threads cares about the order in which two values are written, the other side must be equally concerned about the order in which those values are read.“

**Comment by jaykrell** [comment]
The article and its link go into further detail, on the pairing thing, and show examples that resemble this pattern, the racy case w/o read barrier and fixes with read barrier or data depenency. 

But just merging the variables works, and then need no barrier at all, assuming it is ok to run the check more than once, and that it will always yield the same answer. 

But still caching the value to limit the checks.

**Comment by vargaz** [comment]
@monojenkins build failed

**Comment by duncanmak** [comment]
We should backport this to `2019-02` and `2019-04`, since #13446 was also backported to those branches.

**Comment by lewurm** [comment]
@monojenkins backport 2019-02
@monojenkins backport 2019-04

**Comment by jaykrell** [comment]
Eventual solution: FullAOT?

**Comment by chamons** [comment]
❤️ this solution - No need for an mmp patch now @akoeplinger (did we land one in the integration branch?)

**Comment by akoeplinger** [comment]
@chamons yes we did and it's in xamarin-macios master now so you'll need to revert it there: https://github.com/xamarin/xamarin-macios/blob/81d6225eb7a377fa49dd98aa324f03e749a161dd/runtime/launcher.m#L618

**Comment by chamons** [comment]
https://github.com/xamarin/xamarin-macios/pull/6051

File: mono/utils/mono-mmap.c [hunk 1]
```diff
@@ -287,10 +287,21 @@ mono_valloc (void *addr, size_t length, int flags, MonoMemAccountType type)
 #endif
 
 #if defined(__APPLE__) && defined(MAP_JIT)
```
Comment (jaykrell) [inline mono/utils/mono-mmap.c]: `#ifdef MAP_JIT` here precludes the ability to build on older SDK and run on newer OS (with hardened runtime).
It'd be good to `#define` it ourselves to allow for that. Copy the SDK header content. It is a little gross but the benefits are good and it is not unusual practice.

**Pull request merged**